watertap/core/plugins/tests/test_solvers.py

```diff
@@ -0,0 +1,148 @@
+###############################################################################
+# WaterTAP Copyright (c) 2021, The Regents of the University of California,
+# through Lawrence Berkeley National Laboratory, Oak Ridge National
+# Laboratory, National Renewable Energy Laboratory, and National Energy
+# Technology Laboratory (subject to receipt of any required approvals from
+# the U.S. Dept. of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/watertap-org/watertap/"
+#
+###############################################################################
+
+import pytest
+import pyomo.environ as pyo
+
+from pyomo.common.errors import ApplicationError
+from idaes.core.util.scaling import (set_scaling_factor, get_scaling_factor,
+        constraints_with_scale_factor_generator, unscaled_constraints_generator)
+from idaes.core.util import get_solver
+from watertap.core.plugins.solvers import IpoptWaterTAP
+
+class TestIpoptWaterTAP:
+    @pytest.fixture(scope="class")
+    def m(self):
+        m = pyo.ConcreteModel()
+        m.a = pyo.Var(initialize=1)
+        m.b = b = pyo.Block()
+        b.a = pyo.Var([1,2])
+
+        m.c = pyo.Constraint(expr=(0, 1./(m.a**2) ,1))
+        b.c = pyo.Constraint([1,2], rule= lambda b,i: (i-1, b.a[i], i+2))
+        b.d = pyo.Constraint(expr=b.a[1]**4 + b.a[2]**4 <= 4)
+        set_scaling_factor(b.d, 1e6)
```

Comment:
What's the difference between using this and `constraint_scaling_transform`? As I look at most of our `calculate_scaling_factors` methods in unit models, we usually use `constraint_scaling_transform` to apply scaling factors to constraints (or at least I thought). I was originally under the impression that `set_scaling_factor` only applied to variables which I'm now seeing was an incorrect interpretation.

Reply:
These have an interesting interplay: `constraint_scaling_transform` modifies the Pyomo constraint directly, such that after the transform, this is what the "unscaled" constraint looks like to Ipopt. You can also apply a constraint scaling factor, which will be applied by Ipopt if `nlp_scaling_method=user-scaling` for its interior point method.

In either case, whether specifying a constraint scaling factor and using `user-scaling` **or** just using the `gradient-based` scaling, the transformed constraint will be potentially re-scaled again for the solve, but critically its unscaled (to Ipopt) tolerance will be adjusted by the `constraint_scaling_transform`.